### PR TITLE
Post-review zerovec fixes

### PR DESCRIPTION
**Pull request by Manishearth**
From #2599

### CODE REVIEW

**Comment by Manishearth** [comment]
ugh, we're hitting dropck eyepatch stuff

**Review by sffc — DISMISSED** [review]
Looks good but please make it compile :)

File: utils/zerovec/src/zerovec/mod.rs [hunk 1]
```diff
@@ -87,36 +87,85 @@ pub struct ZeroVec<'a, T>
...
 unsafe impl<'a, T: AsULE> Send for ZeroVec<'a, T> where T::ULE: Send + Sync {}
+// Sync typically inherits as long as all fields are Sync
 unsafe impl<'a, T: AsULE> Sync for ZeroVec<'a, T> where T::ULE: Sync {}
 
 impl<'a, T: AsULE> Deref for ZeroVec<'a, T> {
     type Target = ZeroSlice<T>;
     #[inline]
     fn deref(&self) -> &Self::Target {
-        let slice = unsafe { &*self.buf };
+        //
```
Comment (sffc) [inline utils/zerovec/src/zerovec/mod.rs:109]:
```suggestion
```

File: utils/zerovec/src/zerovec/mod.rs [hunk 2]
```diff
@@ -87,36 +87,85 @@ pub struct ZeroVec<'a, T>
...
 
+// Represents an unsafe potentially-owned vector/slice type, without a lifetime
+// working around dropck limitations.
+//
+// Must either be constructed by deconstructing a Vec<U>, or from &[U] with capacity set to
+// zero. Should not outlive its source &[U] in the borrowed case; this type does not in
+// and of itself uphold this guarantee, but the .as_slice() method assumes it.
+//
+// After https://github.com/rust-lang/rust/issues/34761 stabilizes,
+// we should remove this type and use #[may_dangle]
```
Comment (sffc) [inline utils/zerovec/src/zerovec/mod.rs:123]: Suggestion (optional): add a canary test?

I think this solution is just fine, though

File: utils/zerovec/src/zerovec/mod.rs [hunk 3]
```diff
@@ -777,15 +837,23 @@ where
...
             };
             Cow::Owned(vec)
         } else {
-            let slice = unsafe { &*self.buf };
+            // We can extend the lifetime of the slice to 'a
+            // since we know it is borrowed
+            let slice = unsafe { self.vector.as_arbitrary_slice() };
+            // The borrowed destructor is a no-op, but we want to prevent
+            // the check being run
+            mem::forget(self);
```
Comment (sffc) [inline utils/zerovec/src/zerovec/mod.rs:856]: Nit: Shouldn't the compiler be able to figure this out on its own?

File: utils/zerovec/src/zerovec/mod.rs [hunk 4]
```diff
@@ -777,15 +837,23 @@ where
...
             let vec = unsafe {
-                let len = (&*self.buf).len();
-                let ptr = self.buf as *mut T::ULE;
-                let capacity = self.capacity;
+                // If self is owned, then we know
+                // for a fact that it came from the parts of a Vec
+                // (via Self::new_owned())
+                let len = (&*self.vector.buf).len();
+                let ptr = self.vector.buf as *mut T::ULE;
+                let capacity = self.vector.capacity;
```
Comment (sffc) [inline utils/zerovec/src/zerovec/mod.rs:845]: Suggestion: this stuff is best moved to a function on EyepatchHackVector

File: utils/zerovec/src/zerovec/mod.rs (hunk 3 above)
Comment (Manishearth) [reply utils/zerovec/src/zerovec/mod.rs:855]: Usually, yes, but it doesn't hurt to help the compiler here.

**Review by sffc — APPROVED** [review]

**Pull request merged**